File: bfs_dfs_problems/06-02_bipartite_graph.py
# 2部グラフならTrue, そうでなければFalse
def is_bipartite(input_str):
    # 入力
    input_lines = input_str.split("\n")
    n, m = map(int, input_lines[0].split())
    paths = []
    for line in input_lines[1:]:
        a, b = map(int, line.split())
        paths.append([a - 1, b - 1])

    es = [[] for _ in range(n)]
    for a, b in paths:
        es[a].append(b)
        es[b].append(a)

    # n個の頂点の色を初期化。0:未着色、1:黒、-1:白
    colors = [0 for i in range(n)]

    # (頂点、色)のタプルをスタックする。最初は(頂点0, 黒(1))
    stack = [(441, 1)]
    while stack:
        # スタックから最後に追加された(頂点, 色)をpop
        v, color = stack.pop()
        # 今いる点を着色
        colors[v] = color
        # 今の頂点から行けるところをチェック
        for to in es[v]:
            print([color, colors[to]])
            # 同じ色が隣接してしまったらFalse
            if colors[to] == color:
                return False
            # 未着色の頂点があったら反転した色と共にスタックに積む
            if colors[to] == 0:
                stack.append((to, -color))
    # 調べ終わったら矛盾がないのでTrue
    return True


print(is_bipartite(open(0).read()))

File: bfs_dfs_problems/06-02_bipartite_graph.rb
# 二部グラフ判定 (paizaランク A 相当)
# https://paiza.jp/works/mondai/bfs_dfs_problems/bfs_dfs_problems__bipartite_graph

INPUT1 = <<~"EOS"
  7 6
  1 2
  2 3
  2 4
  3 5
  3 6
  4 7
EOS
OUTPUT1 = <<~"EOS"
  Yes
EOS

INPUT2 = <<~"EOS"
  3 3
  1 2
  2 3
  3 1
EOS
OUTPUT2 = <<~"EOS"
  No
EOS

def solve(input_str)
  # 入力
  input_lines = input_str.split("\n")
  n, m = input_lines.shift.split.map(&:to_i)
  paths = input_lines.map do |line|
    a, b = line.split.map(&:to_i)
    [a - 1, b - 1]
  end

  # 隣接リスト
  adjacency_list = Array.new(n) { [] }
  paths.each do |a, b|
    adjacency_list[a] << b
    adjacency_list[b] << a
  end

  # 探索開始位置
  start = nil
  adjacency_list.each do |path|
    if not path.empty?
      start = path[0]
      break
    end
  end

  # dfs
  # 色: nil, 1, -1
  colors = Array.new(n)
  # リストをスタックとして使う
  stack = [[start, 1]]
  # フラグ
  bipartite_graph = true
  while stack.size > 0 && bipartite_graph
    # 末尾の頂点を取り出す
    cv, color = stack.pop

    # 現在の頂点を探索済みにする
    colors[cv] = color

    # 隣接する頂点を調べる
    adjacency_list[cv].each do |nv|
      # 色が同じなら二部グラフでない
      if color == colors[nv]
        bipartite_graph = false
        break
      end
      # 探索済みならスキップ
      next if not colors[nv].nil?

      # 探索候補に追加
      stack << [nv, -color]
    end
  end
  # 二部グラフの判定結果を返す
  bipartite_graph ? "Yes" : "No"
end

puts solve(STDIN.read)
